plugins/home/src/extensions.tsx

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { lazy, Suspense } from 'react';
+import { IconButton } from '@material-ui/core';
+import SettingsIcon from '@material-ui/icons/Settings';
+import { InfoCard } from '@backstage/core-components';
+import { SettingsModal } from './components';
+import { createReactExtension, useApp } from '@backstage/core-plugin-api';
+
+const lazyLoadedComponent = (
+  component: () => Promise<(props: any) => JSX.Element>,
+) => lazy(() => component().then(inner => ({ default: inner })));
+
+export function createHomePageComponentExtension({
+  title,
+  component,
+  actions,
+  contextProvider,
+  settings,
+}: {
+  title: string;
+  component: () => Promise<(props: any) => JSX.Element>;
+  actions?: () => Promise<(props: any) => JSX.Element>;
+  contextProvider?: () => Promise<(props: any) => JSX.Element>;
+  settings?: () => Promise<(props: any) => JSX.Element>;
+}) {
+  const Content = lazyLoadedComponent(component);
+  const Actions = actions ? lazyLoadedComponent(actions) : () => null;
+  const Settings = settings ? lazyLoadedComponent(settings) : () => null;
+  const ContextProvider = contextProvider
+    ? lazyLoadedComponent(contextProvider)
+    : () => null;
+
+  const HomePageExtension = () => {
```

Comment:
Separate conversation, but same spot!  **Props**

It seems like there will very likely be cases where we want to provide **App Integrators** (as opposed to end-users) a degree of configurability.  As a naive example, imagine a `<HomepageSearchBar />`...  I might want to make the placeholder text of the search bar configurable to the App integrator, e.g.

```tsx
<HomepageSearchBar placeholder="Search Acme Borg" />
```

It could make sense to open this up to passing props.  ...But then, how would you know which sub-component to pass the props to?  In the case above, it would probably go to the `component`, but there might be cases where you'd want to wire it up to the `context`, or `actions`, or even `settings`?

Reply:
That's a great point. We definitely want it so that the integrator can provide props down to the HomePage Components. Another way of doing it, other than the MUI suggestion, would be to let the HomePage Component renderer be responsible for passing childprops to the component context. This gets around the problem where the integrator has to which building block of the HomePage Component will accept which props, and potentially also pass the same prop multiple times.

However, this also means that you're required to register a context to accept props, which isn't optimal since the context isn't strictly necessary otherwise. But perhaps that's an acceptable price to pay to create a nicer interface.

Either way we go, this does put some strain on the way you have to document the HomePage Components, since the integrator would have to know which props they can pass.

Reply:
This has been implemented in 0cce460, where component developers can define custom props they will accept, which the component renderer is responsible for passing to the context.